design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to run [CockroachDB] in a Kubernetes
+cluster. Assuming you have installed Rook and the corresponding CockroachDB
+operator into a cluster, you should be able to dynamically provision a
+CockroachDB cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a non-portable Rook CockroachDB `ClusterClass`
+
+```yaml
+apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+kind: ClusterClass
+metadata:
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+specTemplate:
+  clusterSelector:
+    matchLabels:
+        app: my-cool-cluster
+  # full documentation on all available settings can be found at:
+  # https://rook.io/docs/rook/master/cockroachdb-cluster-crd.html
+  scope:
+    nodeCount: 3
+    # You can only have one PersistentVolumeClaim in this list!
+    volumeClaimTemplates:
+    - metadata:
+        name: rook-cockroachdb-data
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        # Uncomment and specify your StorageClass, otherwise
+        # the cluster admin defined default StorageClass will be used.
+        #storageClassName: "your-cluster-storageclass"
+        resources:
+          requests:
+            storage: "1Gi"
+  network:
+    ports:
+    - name: http
+      port: 8080
+    - name: grpc
+      port: 26257
+  secure: false
+  cachePercent: 25
+  maxSQLMemoryPercent: 25
+  # A key/value list of annotations
+  annotations:
+  #  key: value
+```
+
+2. Create a portable `PostgreSQLInstanceClass`
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstanceClass
+metadata:
+  name: postgresql-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ClusterClass
+  apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+```
+
+3. Create a `PostgreSQLInstance` claim
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: app-postgresql
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: postgresql-standard
+  writeConnectionSecretToRef:
+    name: postgresqlconn
+  engineVersion: "9.6"
+```
+
+On creation of the `PostgreSQLInstance` claim, the claim controller would
+configure and create a `Cluster` (`cockroachdb.rook.crossplane.io/v1alpha1`)
+instance. Then, the managed controller would use the Crossplane `Cluster`
+instance to create a `Cluster` (`cockroachdb.rook.io/v1alpha1`) in the target
+Kubernetes cluster where Rook and the CockroachDB operator are installed. This
```

Comment:
I'm thinking that for on-premises scenarios, having "import" or "bring your own" functionality would be really useful, i.e. an existing cluster already running on-premises being imported into Crossplane so that it can be used as a target cluster for scheduling.

Being able to deploy resources that are managed by operators to target k8s clusters is still useful in managed k8s clusters that we first spin up in a cloud provider using Crossplane, but it may be more useful for true on-premises scenarios instead.

This is tracked in https://github.com/crossplaneio/crossplane-runtime/issues/22.